### PR TITLE
[Admin][Product] Sort products within taxon

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | yes
| Related tickets | -
| License         | MIT

### CODE REVIEW

File: features/product/managing_products/adding_products_with_position_and_taxon.feature [hunk 1]
```diff
@@ -0,0 +1,113 @@
+@managing_products
+Feature: Adding a new product
```
Comment (TheMadeleine) [inline features/product/managing_products/adding_products_with_position_and_taxon.feature]: This feature title  and body do not describe well enough what we are checking here.

```
Feature: Adding a new product with a position within a taxon
In order to organize products within taxons in my shop
As an Administrator
I want to assign positions in taxons to products when I create them
```

I would suggest sth like that.

File: features/product/managing_products/adding_products_with_position_and_taxon.feature [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        Given I want to create a new simple product
+        When I specify its code as "OLD_PUG"
+        And I name it "Old Pug" in "English (United States)"
+        And I set its price to "$10.00"
+        And I assign its to "Soft Toys" taxon
+        And I set its position as "-12"
+        And I add it
+        Then I should be notified that it has been successfully created
+        And the product "Old Pug" should appear in the shop
+        And its position should be 0 in taxon "Soft Toys"
```
Comment (TheMadeleine) [inline features/product/managing_products/adding_products_with_position_and_taxon.feature]: ``its position should be 0 in the "Soft Toys" taxon``

File: features/product/managing_products/adding_products_with_position_and_taxon.feature [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        And the store classifies its products as "Soft Toys" and "Dolls"
+        And I am logged in as an administrator
+
+    @todo
+    Scenario: Adding a new simple product with wrong position
+        Given I want to create a new simple product
+        When I specify its code as "OLD_PUG"
+        And I name it "Old Pug" in "English (United States)"
+        And I set its price to "$10.00"
+        And I assign its to "Soft Toys" taxon
```
Comment (TheMadeleine) [inline features/product/managing_products/adding_products_with_position_and_taxon.feature]: ``I assign it to the "Soft Toys" taxon``

File: features/product/managing_products/adding_products_with_position_and_taxon.feature [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        When I specify its code as "OLD_PUG"
+        And I name it "Old Pug" in "English (United States)"
+        And I set its price to "$10.00"
+        And I add it
+        Then I should be notified that it has been successfully created
+        And the product "Old Pug" should appear in the shop
+        And its position should be 2 in taxon "Soft Toys"
+
+    @todo
+    Scenario: Adding a new simple product in many taxons with different positions
```
Comment (TheMadeleine) [inline features/product/managing_products/adding_products_with_position_and_taxon.feature]: ``Adding a new simple product with different positions in many taxons``

File: features/product/managing_products/adding_products_with_position_and_taxon.feature [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        And I add it
+        Then I should be notified that it has been successfully created
+        And the product "Old Pug" should appear in the shop
+        And its position should be 0 in taxon "Soft Toys"
+
+    @todo
+    Scenario: Adding a new simple product without position
+        And the store has a product "Young pug"
+        And this product belongs to "Soft Toys"
+        And this product is at position 1 at this taxon
```
Comment (TheMadeleine) [inline features/product/managing_products/adding_products_with_position_and_taxon.feature]: ``this product is at position 1 in this taxon``

File: features/product/managing_products/sorting_products_within_a_taxon_by_position.feature [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  @todo
+  Scenario: Products are sorted by position in ascending order by default
+    When I am browsing products from taxon "Soft Toys"
+    Then I should see 3 products in the list
+    And the first product on the list should have name "Old Pug"
+    And the last product on the list should have name "Small Pug"
+
+  @todo
+  Scenario: Shipping method added at no position is added as the last one
```
Comment (TheMadeleine) [inline features/product/managing_products/sorting_products_within_a_taxon_by_position.feature]: ``A shipping method added without position is added as the last one``

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 7]
```diff
@@ -445,6 +445,26 @@ public function iShouldSeeTheProductAssociationWithProducts($productAssociationN
     }
 
     /**
+     * @Then /^average rating of (product "[^"]+") should be (\d+)$/
+     */
+    public function thisProductAverageRatingShouldBe(ProductInterface $product, $averageRating)
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: I think this does not belong to this PR?

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 8]
```diff
@@ -445,6 +445,26 @@ public function iShouldSeeTheProductAssociationWithProducts($productAssociationN
...
+        $this->iShouldSeeAsItsAverageRating($averageRating);
+    }
+
+    /**
+     * @Then they should have order like :firstProductName, :secondProductName and :thirdProductName
+     */
+    public function theyShouldHaveOrderLikeAnd(...$productNames)
+    {
+        Assert::true(
+            $this->taxonShowPage->hasProductsInOrder($productNames),
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: 👍

File: src/Sylius/Behat/Page/Admin/Product/IndexPerTaxonPageInterface.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    public function hasProductsInOrder(array $productNames);
+
+    /**
+     * @param string $productName
+     * 
+     * @param int $position
+     */
+    public function setPositionOfProduct($productName, $position);
+
+    public function saveConfiguration();
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Page/Admin/Product/IndexPerTaxonPageInterface.php]: savePositions :)

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_taxon.yml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+sylius_admin_product_taxons_update:
+    path: product-taxons/update
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_taxon.yml]: Should start with ``/``.

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_taxon.yml [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+    methods: [GET, PUT]
+    defaults:
+        _controller: sylius.controller.productTaxon:updatePositionsAction
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/product_taxon.yml]: ``sylius.controller.product_taxon`` should be the name.

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/ProductTaxon.orm.xml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+-->
+
+<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping"
+                  xmlns:gedmo="http://gediminasm.org/schemas/orm/doctrine-extensions-mapping"
+                  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                  xsi:schemaLocation="http://doctrine-project.org/schemas/orm/doctrine-mapping
+                                      http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd">
+
+    <mapped-superclass name="Sylius\Component\Core\Model\ProductTaxon" table="sylius_product_taxon">
+        <id name="id" column="id" type="integer">
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/ProductTaxon.orm.xml]: Do we need id column here?

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php (hunk 7 above)
Comment (tuka217) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: Yes, but this step was under private method, so I moved it before `assertIsProductIsInAssociation` and after that I added my new step. Therefore this lines are marks as new step.

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/ProductTaxon.orm.xml (hunk 12 above)
Comment (tuka217) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/ProductTaxon.orm.xml]: Yups

**Comment by michalmarcinkowski** [comment]
Travis was stalled, so I needed to trigger the hook once again.

**Comment by pjedrzejewski** [comment]
Thank you Ania, nice work! :)

**Comment by tchapi** [comment]
How do we find if a product has a taxon if we only know the slug of the taxon now ?

Suppose I have a filtered `/products/{taxonSlug}` route, and in my controller :

```php
[...]
$taxon = $this->get('sylius.repository.taxon')->findOneBySlug($taxonSlug);

// Do I need to iterate over each productTaxon of each product ??
foreach ($products as $product) {
  foreach($product->getProductTaxons() as $productTaxon) {
    if ($productTaxon()->getTaxon() == $taxon) {
       // filter or do whatever
    }
  }
}
```
This seems a bit overkill, no ? Did I miss something ?

**Comment by michalmarcinkowski** [comment]
How about:
`$this->get('sylius.repository.product')->findByTaxonSlug($taxonSlug);`

```
return $this->createQueryBuilder('o')
            ->innerJoin('o.taxons', 'taxon')
            ->andWhere('taxon.slug = :slug')
            ->setParameter('slug', $taxonSlug)
            ->getQuery()
            ->getResult()
        ;
```

**Comment by michalmarcinkowski** [comment]
Anyway the previous implementation would look almost the same as yours now:
```
foreach ($products as $product) {
  foreach($product->getTaxons() as $productTaxon) {
    if ($productTaxon == $taxon) {
       // filter or do whatever
    }
  }
}
```

**Comment by tchapi** [comment]
The for loop looks the same but doctrine hydrates thing differentely between the two, and performance can get tricky. I'll go with the Repository method I guess
